src/hazelcore/extractions/ocaml/OCamlExtraction_Exp.re

```diff
@@ -0,0 +1,270 @@
+
+//type extract_result = option((ocaml_exp, HTyp.t));
+type t = (ocaml_exp, HTyp.t)
+and ocaml_exp = string;
+
+/* Theorem: If extract(ctx, d) = Some(ce, ty) and ctx |- d : ty then ce :_ocaml Typ.extract(ty) */
+
+//TODO: replace == with HTyp.consistent
+let rec extract = (~ctx: Contexts.t, ~de: DHExp.t): t =>
+  switch (de) {
+  | EmptyHole(_) => failwith("Exp: Empty Hole")
+  | NonEmptyHole(_) => failwith("Exp: Non-Empty Hole")
+  // used for variables that are keywords in Hazel, like "let" and "case"
+  | Keyword(_) => failwith("Exp: Incomplete Program (Keyword)")
+  // an expression with a free variable cannot be extracted to OCaml because free variables cause type errors in OCaml
+  | FreeVar(_, _, _, _) => failwith("Exp: FreeVar is not allowed")
+  | BoundVar(x) =>
+    //directly lookup type in environment
+    let typ = VarCtx.lookup(Contexts.gamma(ctx), x);
+    switch (typ) {
+    | None => failwith("Exp: BoundVar " ++ x ++ " Not Found")
+    // if the hole type, we don't print the type
+    | Some(t) => (x, t)
+    // if (t == Hole) {
+    //   ("(" ++ x ++ ")", t);
+    // } else {
+    //   ("(" ++ x ++ " : " ++ Typ.extract(~t) ++ ")", t);
+    // }
+    };
+  | Let(dp, de1, de2) =>
+    switch (dp) {
+    | Var(x) =>
+      let ocaml_dp = OCamlExtraction_Pat.extract(~dp);
+      let ocaml_de1 = extract(~ctx, ~de=de1);
+      let ctx' = Contexts.extend_gamma(ctx, (x, snd(ocaml_de1)));
+      let ocaml_de2 = extract(~ctx=ctx', ~de=de2);
+      // use the "let .. = ..;;" format
+      // if it's a recursive function, it's format of let ... = fix ..., we change it to "let rec"
+      //, because the recursive function should be declared by "rec" only in ocaml
+      switch (de1) {
+      | FixF(_, _, _) => (
+          "let rec "
+          ++ ocaml_dp
+          ++ " : "
+          ++ OCamlExtraction_Typ.extract(~t=snd(ocaml_de1))
+          ++ " = "
+          ++ fst(ocaml_de1)
+          ++ ";;\n"
+          ++ fst(ocaml_de2),
+          snd(ocaml_de2),
+          // fst(ocaml_de1) ++ fst(ocaml_de2),
+          // snd(ocaml_de2),
+        )
+      | _ => (
+          "let "
+          ++ ocaml_dp
+          ++ " : "
+          ++ OCamlExtraction_Typ.extract(~t=snd(ocaml_de1))
+          ++ " = "
+          ++ fst(ocaml_de1)
+          ++ ";;\n"
+          ++ fst(ocaml_de2),
+          snd(ocaml_de2),
+        )
+      };
+    | _ => failwith("Exp: Let, not a variable pattern")
+    }
+  | FixF(x, ht, de) =>
+    //it's the case of fixpoint/ recursive functions
+    //Hazel programs will only ever result in internal expressions containing let ... = fix ...
+    //let rec f: some type = fun x -> f ... (is reasonable)
+    // x should be already inserted into context
+
+    //let ocaml_ht = OCamlExtraction_Typ.extract(~t=ht);
+    let ctx' = Contexts.extend_gamma(ctx, (x, ht));
+    let ocaml_de = extract(~ctx=ctx', ~de);
+    (
+      // "let rec " ++ x ++ " : " ++ ocaml_ht ++ " = " ++ fst(ocaml_de) ++ ";;\n",
+      fst(ocaml_de),
```

Comment:
instead of using fst, use pattern matching in the let above

Reply:
(happens a few other times too, will just mention it once here)

Reply:
May I ask why we don't use fst and snd? I think fst and snd may make the code looks a lot better, because we will call then multiple times in a block...

Reply:
Oh, I just rewrote, it actually looks better in pattern matching.